### PR TITLE
Matching Revamp

### PR DESCRIPTION
**Pull request by SansPapyrus683**
All solutions have been tested.

### CODE REVIEW

File: solutions/ac-matching.mdx [hunk 1]
```diff
@@ -35,115 +36,84 @@ considered as a single matching involving zero pairs).
...
-int dp[1 << 21];
+const int MOD = 1e9 + 7;
+const int MAX_N = 21;
 
-int N;
+bool compat[MAX_N][MAX_N];
+int dp[1 << MAX_N];
 
 int main() {
-	cin.tie(0)->sync_with_stdio(0);
```
Comment (thecodingwizard) [inline solutions/ac-matching.mdx:44]: why? D:

**Comment by thecodingwizard** [comment]
Didn't look too closely, but seems pretty good. Thanks!

File: solutions/ac-matching.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply solutions/ac-matching.mdx:44]: isn't necessary

Comment (thecodingwizard) [reply solutions/ac-matching.mdx:44]: so? might as well include it as part of every C++ program :c

(ah well idrc, if I'm the only one who feels this way then whatever)

Comment (HarryWangATX) [reply solutions/ac-matching.mdx:44]: doesnt it speed io up?

Comment (nwatx) [reply solutions/ac-matching.mdx:44]: yes, but n is like <= 20

> doesnt it speed io up?

Comment (HarryWangATX) [reply solutions/ac-matching.mdx:44]: oh yeaa xd. ig it doesn't rly matter if we put this, but it does look cool 😆

Comment (nwatx) [reply solutions/ac-matching.mdx:44]: yeah same I am quite impartial on this :laughing:

**Review by dongliu0426 — APPROVED** [review]
```diff
- cin.tie(0)->sync_with_stdio(0);
```
questionable change...

**Comment by SansPapyrus683** [comment]
it still get's ac'd, so i don't really see a reason for it in the first place

**Comment by HarryWangATX** [comment]
i guess.

**Pull request merged**